crates/nu-protocol/src/value/record.rs

```diff
@@ -26,24 +26,23 @@ impl Record {
         }
     }
 
-    /// Constructor that checks that `cols` and `vals` are of the same length.
+    /// Unsafely create a [`Record`] from a `Vec` of columns and a `Vec` of [`Value`]s
     ///
-    /// WARNING! Panics with assertion failure if cols and vals have different length!
-    /// Should be used only when the same lengths are guaranteed!
-    ///
-    /// For perf reasons does not validate the rest of the record assumptions.
+    /// For perf reasons, this will not validate the rest of the record assumptions:
     /// - unique keys
-    pub fn from_raw_cols_vals_unchecked(cols: Vec<String>, vals: Vec<Value>) -> Self {
-        assert_eq!(cols.len(), vals.len());
-
+    ///
+    /// # Safety
+    /// This function should only be used if `cols` and `vals` have the same length.
+    pub unsafe fn from_raw_cols_vals_unchecked(cols: Vec<String>, vals: Vec<Value>) -> Self {
+        debug_assert_eq!(cols.len(), vals.len());
```

Comment:
I am not really sure if making this just a debug assertion accelerates our goal of getting rid of bad constructions.

They are bugs that should absolutely be removed. So they should panic now and not fly in our face much later.

Maybe I kind of disagree with the setup of `from_raw_cols_vals` in #11718 by @kubouch 

It should either blow up in your face and be fixed or be designed in such a way that you can not provide illegal values.
Length is one side of the equation but we should also move more towards intefaces that enforce unique keys. For the latter the checking is more costly (I would wager that the gain of eliding that length check here is minor compared to potential pain)

Reply:
> They are bugs that should absolutely be removed. So they should panic now and not fly in our face much later.

`debug_assert_eq!` does panic in debug mode. Panicing in release as well could help us find these bugs more easily but there's no guarantee all will be caught. The problem with panicing in release mode is that it could lock people out of their systems. Unless we add some sort of panic recovery, I don't think we should be panicing at all in release. Then again, mismatched lengths can cause other code to panic anyways, so this might be a minor point -- but I still think it's better to avoid panics as much as possible. To this end, I think `from_raw_cols_vals` is preferable. Maybe we could add a different class of errors (i.e., `ShellError::InternalError(InternalError)`) to indicate that an error is a bug and not a user error.

> Length is one side of the equation but we should also move more towards intefaces that enforce unique keys. 

Yep, for sure. This only tackles length for now.

Reply:
The reason I did it how I did it was because I just wanted to fix the `rotate` panic and not fix all the 100 other call sites. So I used the unchecked variant everywhere else, matching the existing functionality with the intent to be fixed later (in this PR).

Reply:
Once the panic recovery is added for the repl, it should be fine to use `assert!`. Again, the problem with panicing in release is that this could prevent people from logging in to their system if nushell always crashes on start. For example, if anything in the user's config files manipulates records, etc, then this could trigger a panic.

Maybe I should just remove `from_raw_cols_vals_unchecked` altogether. There's only two places using it besides the macro.